synapse/storage/schema/main/delta/60/99remove_hidden_devices_from_device_inbox.sql

```diff
@@ -0,0 +1,20 @@
+/* Copyright 2021 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+--- Remove messages to hidden devices from the device_inbox table.
+--- This schould run as the last task, it may take a little bit longer
+--- to finish.
+
+DELETE FROM device_inbox WHERE device_id IN (SELECT device_id FROM devices WHERE hidden = TRUE);
```

Comment:
Yes it will run a long time. Here manually with 110 000 000 rows it took few hours. For small homeservers it would be fine. Do i understand it correctly it will iterate over every row in the table? If it does, it would be useful to add a third pr which removes the sql delta files here and in #10098 and do something like this as background update instead.
```
DELETE FROM device_inbox 
    WHERE device_id IN (SELECT device_id FROM devices WHERE hidden = TRUE)
    OR WHERE device_id NOT IN (SELECT device_id FROM devices);
```

Reply:
And gives the opportunity to test it now, with the slow version.

Reply:
Good point, it's probably best to encapsulate the background job in a separate PR as it's not required for the main functionality of this PR anyhow.

> Do i understand it correctly it will iterate over every row in the table?

The currently active indexes on this table are:

```
Indexes:
    "device_inbox_stream_id" btree (stream_id)
    "device_inbox_user_stream_id" btree (user_id, device_id, stream_id)
```

Because of the lack of `stream_id`, I believe we'd need to iterate over the table.

> And gives the opportunity to test it now, with the slow version.

Sure, people can run the SQL during operation manually if they'd wish, and then we just add a background update separately to clean it out in parallel for larger servers :slightly_smiling_face:

Reply:
A little off topic for here, but I noticed in our database while trying to do a similar cleanup that even selecting the rows to be removed was a really slow operation. Something like `SELECT * FROM device_inbox WHERE device_id = 'foo'` took several minutes to run. We might have to be a little clever in how we clean up this table.

Reply:
Doesn't look like there's an index for just the `user_id`. In `full.sql.postgres` I see

```sql
CREATE INDEX device_inbox_stream_id_user_id ON device_inbox USING btree (stream_id, user_id);
CREATE INDEX device_inbox_user_stream_id ON device_inbox USING btree (user_id, device_id, stream_id);
```